instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/HostIdResourceProvider.java

```diff
@@ -0,0 +1,209 @@
+    this.queryWindowsRegistry = queryWindowsRegistry;
+  }
+
+  @Override
+  public Resource createResource(ConfigProperties config) {
+    OsType osType = getOsType.get();
+    switch (osType) {
+      case WINDOWS:
+        return readWindowsGuid();
+      case LINUX:
+        return readLinuxMachineId();
+      case OTHER:
+        break;
+    }
+    logger.info("Unsupported OS type: " + osType);
+    return Resource.empty();
+  }
+
+  private Resource readLinuxMachineId() {
+    Path path = FileSystems.getDefault().getPath("/etc/machine-id");
+    List<String> lines = machineIdReader.apply(path);
+    if (lines.isEmpty()) {
+      return Resource.empty();
+    }
+    return Resource.create(Attributes.of(ResourceAttributes.HOST_ID, lines.get(0)));
+  }
+
+  private static List<String> readMachineIdFile(Path path) {
+    try {
+      List<String> lines = Files.readAllLines(path);
+      if (lines.isEmpty()) {
+        logger.info("Failed to read /etc/machine-id: empty file");
+      }
+      return lines;
+    } catch (IOException e) {
+      logger.log(Level.INFO, "Failed to read /etc/machine-id", e);
+      return Collections.emptyList();
+    }
+  }
+
+  private static OsType getOsType() {
+    // see
+    // https://github.com/apache/commons-lang/blob/master/src/main/java/org/apache/commons/lang3/SystemUtils.java
+    // for values
+    String osName = System.getProperty("os.name");
+    if (osName == null) {
+      return OsType.OTHER;
+    }
+    if (osName.startsWith("Windows")) {
+      return OsType.WINDOWS;
+    }
+    if (osName.toLowerCase(Locale.ROOT).equals("linux")) {
+      return OsType.LINUX;
+    }
+    return OsType.OTHER;
+  }
+
+  private Resource readWindowsGuid() {
+    try {
+      ExecResult execResult = queryWindowsRegistry.get();
+
+      if (execResult.exitCode != 0) {
+        logger.info(
+            "Failed to read Windows registry. Exit code: "
+                + execResult.exitCode
+                + " Output: "
+                + String.join("\n", execResult.lines));
+        return Resource.empty();
+      }
+
+      for (String line : execResult.lines) {
+        if (line.contains("MachineGuid")) {
+          String[] parts = line.trim().split("\\s+");
+          if (parts.length == 3) {
+            return Resource.create(Attributes.of(ResourceAttributes.HOST_ID, parts[2]));
+          }
+        }
+      }
+      logger.info(
+          "Failed to read Windows registry: No MachineGuid found in output: " + execResult.lines);
+      return Resource.empty();
+    } catch (RuntimeException e) {
+      logger.log(Level.INFO, "Failed to read Windows registry", e);
+      return Resource.empty();
+    }
+  }
+
+  private static ExecResult queryWindowsRegistry() {
+    try {
+      ProcessBuilder processBuilder = new ProcessBuilder("cmd", "/c", REGISTRY_QUERY);
+      processBuilder.redirectErrorStream(true);
+      Process process = processBuilder.start();
+
+      if (!process.waitFor(2, TimeUnit.SECONDS)) {
```

Comment:
If you don't want to create a separate thread to consume the process output then you could consume it on the main thread providing that you have redirected the error stream, though then you could use the timeout.
@trask do you have any guidance on how we should execute external commands. Should we use a timeout? If so how long would a reasonable timeout be? Any advice on handling the process output?

Reply:
I found an easy way to avoid this problem - redirect the output to a file - where we don't have this issue.

Reply:
This file is created in a temporary dir, which should always be available - or you have bigger issues than not detecting a resource AFAIK.

Reply:
> OK, I looked what other people do: https://github.com/quarkusio/quarkus/blob/main/core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildRunner.java#L29-L37

As mentioned before doing this is fine as long as the invoked process does not produce too much output. I guess the preference for doing it correctly vs cutting corners like that comes from whether you have already had the pleasure of debugging issues where the external process returns a bit more output than expected.

Reply:
> what do you do when you can't find a writable location?

this is a good point, read-only file systems are becoming more popular for security reasons

Reply:
> Gave it another try using https://stackoverflow.com/questions/8149828/read-the-output-from-java-exec
> 
> This should be safe to use even for larger outputs without requiring an additional thread, because we read both stdout and stderr on the calling thread.

where do you put the timeout in this case to guard against the underlying process taking a long time to respond?

Reply:
> where do you put the timeout in this case to guard against the underlying process taking a long time to respond?

I think it is fine to omit the timeout until there is evidence that it is needed. Using short timeouts isn't ideal, if it for some reason (perhaps a load spike) happens to timeout the `container.id` would be missing and it would be hard to tell why this happened.

Reply:
I think doing something like that is an overkill.

Reply:
And I'm fine with that - I was just happy that I discovered it :smile: